src/platform/plugins/shared/workflows_extensions/server/steps/data/data_regex_extract_step.ts

```diff
@@ -0,0 +1,120 @@
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { createRegex, validateInputLength, validateSourceInput } from './regex_utils';
+import { dataRegexExtractStepCommonDefinition } from '../../../common/steps/data';
+import { createServerStepDefinition } from '../../step_registry/types';
+
+function extractFieldsFromMatch(
+  match: RegExpMatchArray,
+  fields: Record<string, string>
+): Record<string, unknown> {
+  const extractedFields: Record<string, unknown> = {};
+
+  for (const [fieldName, groupRef] of Object.entries(fields)) {
+    if (groupRef.startsWith('$')) {
+      const index = parseInt(groupRef.slice(1), 10);
+      extractedFields[fieldName] = isNaN(index) ? null : match[index] ?? null;
+    } else {
+      extractedFields[fieldName] = match.groups?.[groupRef] ?? null;
+    }
+  }
+
+  return extractedFields;
+}
+
+export const dataRegexExtractStepDefinition = createServerStepDefinition({
+  ...dataRegexExtractStepCommonDefinition,
+  handler: async (context) => {
+    try {
+      const source = context.contextManager.renderInputTemplate(context.config.source);
+      const errorIfNoMatch = context.config.errorIfNoMatch || false;
+      const { pattern, fields, flags = '' } = context.input;
+
+      const sourceValidation = validateSourceInput(source, context.logger);
+      if (!sourceValidation.valid) {
+        return { error: sourceValidation.error };
+      }
+      const { isArray } = sourceValidation;
+
+      const sourceArray = isArray
+        ? (source as unknown[])
+        : [source as string | Record<string, unknown>];
+      const shouldReturnObject = !isArray;
+
+      if (sourceArray.length === 0) {
+        context.logger.debug('Input array is empty, returning empty array');
+        return { output: [] };
+      }
+
+      context.logger.debug(
+        `Extracting from ${sourceArray.length} item(s) using pattern: ${pattern} with flags: ${flags}`
+      );
+
+      // Validate the regex pattern with the user-provided flags
+      const regexResult = createRegex(pattern, flags, {
+        error: (message: string, error?: unknown) =>
+          context.logger.error(message, error as Error | undefined),
+      });
+      if ('error' in regexResult) {
+        return { error: regexResult.error };
+      }
+
+      // Remove global flag for matching to ensure match() returns capture groups instead of all matches
+      // When 'g' flag is present, match() returns array of all matches without groups
+      const normalizedFlags = flags?.replace(/g/g, '');
+      const matchRegex = new RegExp(pattern, normalizedFlags);
+
+      const extractedItems: Array<Record<string, unknown> | null> = [];
+      let matchCount = 0;
+
+      for (const item of sourceArray) {
```

Comment:
Noticed in other step impls that we might be able to look for context.abortSignal to allow immediate cancellation (run_stack_monitor.ts:70 // ai_prompt_step.ts:56). Consider adding similar check at the start of each iteration here to prevent processing remaining items after cancellation is requested & ensure workflows can still be cancelled promptly when processing large arrays?